### PR TITLE
Fix replication of SmartGraph edge collection data

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix replication of SmartGraph edge collection data

Fix encoding and decoding of revision ids in replication of SmartGraph edge collection data using the Merkle tree incremental sync protocol. Previously, the encoding of revision ids could be ambiguous under some circumstances.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.10: https://github.com/arangodb/arangodb/pull/17251
  - [x] Backport for 3.9: https://github.com/arangodb/arangodb/pull/17252
  - [x] Backport for 3.8: https://github.com/arangodb/arangodb/pull/17253

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: lib/Basics/HybridLogicalClock.h [hunk 1]
```diff
@@ -123,7 +122,9 @@ class HybridLogicalClock {
   /// the result buffer are returned
   static std::pair<size_t, size_t> encodeTimeStamp(uint64_t t, char* r) {
     size_t pos = 11;
+    r[pos - 1] = '\0';
```
Comment (neunhoef) [inline lib/Basics/HybridLogicalClock.h:125]: I do not understand this. If t is equal to 0, we return the pair 11, 0 anyway. If t is not equal to 0, then this ``\0'` is overwritten in the first run of the loop. Anyway, it also does no harm, as far as I can see.

**Review by neunhoef — APPROVED** [review]
LGTM. See one comment, where I do not understand a line of additional, new code. However, it does not harm either.

File: lib/Basics/HybridLogicalClock.h (hunk 1 above)
Comment (jsteemann) [reply lib/Basics/HybridLogicalClock.h:125]: Ah, true.

Comment (jsteemann) [reply lib/Basics/HybridLogicalClock.h]: Fixed.

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 2]
```diff
@@ -3061,10 +3061,25 @@ bool RestReplicationHandler::prepareRevisionOperation(
...
+  // we first try the parameter "resumeHLC" - if set, this will contain a
+  // timestamp-encoded HLC value
+  std::string const& resumeString =
+      _request->value(StaticStrings::RevisionTreeResumeHLC, found);
   if (found) {
-    ctx.resume = RevisionId::fromString(resumeString);
+    // "resumeHLC" is set. use it
+    uint64_t value = HybridLogicalClock::decodeTimeStamp(resumeString);
+    TRI_ASSERT(value != UINT64_MAX);
+    ctx.resume = RevisionId{value};
```
Comment (mpoeter) [inline arangod/RestHandler/RestReplicationHandler.cpp:3073]: Should we add a method `fromHLC` to Revision (similar `fromString`)?

**Review by mpoeter — APPROVED** [review]

File: arangod/RestHandler/RestReplicationHandler.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/RestHandler/RestReplicationHandler.cpp]: Added.

**Comment by jsteemann** [comment]
Obsoleted by https://github.com/arangodb/arangodb/pull/17310

**Pull request closed without merging**